src/net.cpp

```diff
@@ -3677,27 +3661,61 @@ void clientHandlePacket()
 		// clear follower menu entities.
 		FollowerMenu[clientnum].closeFollowerMenuGUI(true);
 
-		numplayers = 0;
-		assignActions(&map);
-		generatePathMaps();
-		for ( node = map.entities->first; node != nullptr; node = nextnode )
-		{
-			nextnode = node->next;
-			Entity* entity = (Entity*)node->element;
-			if ( entity->flags[NOUPDATE] )
-			{
-				list_RemoveNode(entity->mynode);    // we're anticipating this entity data from server
-			}
-		}
+        // load map file
+		loading = true;
+	    createLevelLoadScreen(5);
+	    std::atomic_bool loading_done {false};
+	    auto loading_task = std::async(std::launch::async, [&loading_done](){
+		    gameplayCustomManager.readFromFile();
+	        updateLoadingScreen(10);
+
+		    int checkMapHash = -1;
+		    int result = physfsLoadMapFile(currentlevel, mapseed, false, &checkMapHash);
+		    if ( checkMapHash == 0 )
+		    {
+			    conductGameChallenges[CONDUCT_MODDED] = 1;
+			    gamemods_disableSteamAchievements = true;
+		    }
+	        updateLoadingScreen(50);
 
-		saveGame();
+		    numplayers = 0;
+		    assignActions(&map);
+	        updateLoadingScreen(55);
+
+		    generatePathMaps();
+	        updateLoadingScreen(80);
+
+            node_t *node, *nextnode;
+		    for ( node = map.entities->first; node != nullptr; node = nextnode )
+		    {
+			    nextnode = node->next;
+			    Entity* entity = (Entity*)node->element;
+			    if ( entity->flags[NOUPDATE] )
+			    {
+				    list_RemoveNode(entity->mynode);    // we're anticipating this entity data from server
+			    }
+		    }
+	        updateLoadingScreen(99);
+
+		    loading_done = true;
+		    return result;
+		});
+	    while (!loading_done)
+	    {
+		    doLoadingScreen();
+		    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+	    }
+	    destroyLoadingScreen();
+		loading = false;
+	    int result = loading_task.get();
 
 		// (special) unlock temple achievement
 		if ( secretlevel && currentlevel == 8 )
 		{
 			steamAchievement("BARONY_ACH_TRICKS_AND_TRAPS");
 		}
 
+		saveGame();
```

Comment:
Should savegame() be before the async task gets started? Perhaps there could be potential crashes in loadMap if you have map data from a mod

Reply:
Good idea.